File: frontend/model/challenge.py
from sqlalchemy import Column, ForeignKey, Text, text
from sqlalchemy.dialects.postgresql import JSONB, UUID

from frontend.extensions import db


class ChallengeSet(db.Model):
    __tablename__ = "challenge_set"

    id = Column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=text("uuid_generate_v4()")
    )
    slug = Column(
        Text(),
        index=True,
        nullable=False
    )
    name = Column(
        Text(),
        nullable=False
    )
    description = Column(
        Text(),
        nullable=False
    )
    version = Column(
        Text(),
        nullable=False
    )


class Challenge(db.Model):
    __tablename__ = "challenge"

    id = Column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=text("uuid_generate_v4()")
    )
    parent = Column(UUID(as_uuid=True),
                    ForeignKey("challenge_set.id", ondelete='CASCADE'),
                    nullable=False)
    slug = Column(
        Text(),
        index=True,
        nullable=False
    )
    name = Column(
        Text(),
        nullable=False
    )
    description = Column(
        Text(),
        nullable=False
    )
    provisioner = Column(
        JSONB(),
        nullable=False
    )
    documentation = Column(
        JSONB(),
        nullable=False
    )

File: operator/src/main/kotlin/org/acasictf/ctf/operator/model/Challenge.kt
package org.acasictf.ctf.operator.model

import kotlinx.serialization.Serializable

@Serializable
data class Challenge(
        val id: String,
        val slug: String,
        val name: String,
        val description: String,
        val provisioner: Provisioner,
        val documentation: List<Document>
)

@Serializable
data class Provisioner(
        val type: String
)

@Serializable
data class Document(
        val name: String,
        val path: String
)
